src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1629,12 +1629,27 @@ private NonReentrantLock StateLock
         ///
         /// This not intended to be the public API, use Document.WithFrozenPartialSemantics() instead.
         /// </summary>
-        public SolutionState WithFrozenPartialCompilationIncludingSpecificDocument(DocumentId documentId, CancellationToken cancellationToken)
+        public SolutionState WithFrozenPartialCompilationIncludingSpecificDocument(DocumentId documentId, ImmutableArray<DocumentId> linkedDocumentIds, CancellationToken cancellationToken)
         {
             try
             {
                 var doc = this.GetRequiredDocumentState(documentId);
                 var tree = doc.GetSyntaxTree(cancellationToken);
+                using var _1 = ArrayBuilder<DocumentState>.GetInstance(out var docBuilder);
+                using var _2 = ArrayBuilder<SyntaxTree>.GetInstance(out var treeBuilder);
+
+                docBuilder.Add(doc);
+                treeBuilder.Add(tree);
+
+                foreach (var linkedDocumentId in linkedDocumentIds)
+                {
+                    var linkedDoc = this.GetRequiredDocumentState(linkedDocumentId);
+                    docBuilder.Add(linkedDoc);
+                    treeBuilder.Add(linkedDoc.GetSyntaxTree(cancellationToken));
+                }
+
+                var linkedDocs = docBuilder.AsImmutable();
+                var linkedTrees = treeBuilder.AsImmutable();
```

Comment:
ToImmutable

Reply:
actually... i don't know why you even need these.  i also don't know why they're called linkedXXX since they also contain the non-linked items.